Let make's default target build the compiler

shindere opened the pull request:
This PR makes it possible to build the OCaml compiler according to
its configuration by simply runnning `make`. There is no need to specify
neither world nor world.opt explicitly, although the two targets
remain available.

This PR also introduces (and starts making use of) the
NATIVE_COMPILER build variable whosse value is true when the native
compiler is enabled and false otherwise.

shindere committed:
Remove the unused BYTECODE_ONLY and NATIVECODE_ONLY make variables

dra27 commented on Makefile:
```diff
@@ -22,17 +22,7 @@ include Makefile.common
-else
-	@echo "Please refer to the instructions in file README.win32.adoc."
-endif
+defaultentry: $(DEFAULT_MAKE_TARGET)
```
The [GNU Coding Standards](https://www.gnu.org/prep/standards/html_node/Standard-Targets.html) recommend this target be called `all`.

At the moment, the `all` target rebuilds the system using whatever is presently in `boot/` (so it's used by `world` after first building the runtime and `bootstrap` after a bootstrap cycle). It seems to me we could safely rename that `all` target and use it here?

xavierleroy replied:
Let me explain the current Makefile, because it's better to understand it before changing it.  

The entries are named from the perspective of a developer (someone who changes the OCaml sources) rather than the perspective of a user (someone who just builds from a source distribution).

`world` is the initial build from a fresh git checkout.  It populates `boot/` with a bytecode interpreter and a standard library, then rebuilds everything (in bytecode at least).

After this initialization, `all` is the "normal" build for a developer: after a change, it rebuilds everything in bytecode, using the existing `boot/`.  For many changes -- all those that require a bootstrap later -- it would be wrong to do `world` and rebuild `boot/`.  This could leave the working copy in an inconsistent state, requiring the changes to be rolled back.

The subtlety above is lost on many core developers, who always blindly do `make world`, then complain that bootstrapping is difficult.  So, maybe changing the meaning of `make all` entirely would just add to the existing confusion in negligible amounts.  Still, there is confusion that needs to be taken into account.  "The GNU Coding Standards recommend this and  that" is not enough.

dra27 replied:
Indeed, but I guess I'm asking for `world` to become `all` and `all` to become something else (like `system`, or `tools`, or `compiler` or something else which doesn't imply *everything*)

xavierleroy replied:
>  I'm asking for world to become all

I understood you're asking for `defaultentry` to become `all`.  (`defaultentry` being `world` or `world.opt` depending on platform.)

My point is that `all` (today) has a well defined meaning for core system developers.  There is also `allopt` that is like `all` but for the native-code parts of the system.  We need to come up with a better story than just rename `defaultentry` to `all` and `all` to "`system`, or `tools`, or `compiler` or something else".

dra27 replied:
Yes, sorry - wrote too quickly as to which target was being renamed. With the aid of a fairly trivial "make cookie" (just `boot/built` or something) could we have it that `make all` the first time does `world` (or `world.opt`) as appropriate but the second time (because of the cookie) does exactly what `make all` does now? `make world` continues to mean the same as it did before, but most of the time it means that just `make` on its own actually does the correct thing, both for a core developer and for an end-user?

dra27 commented on Makefile:
```diff
@@ -22,17 +22,7 @@ include Makefile.common
-	@echo "should work.  But see the file INSTALL for more details."
-else
-	@echo "Please refer to the instructions in file README.win32.adoc."
-endif
```
The error message below has been less relevant since 4.08, since if you forget to run `configure`, then these two `include`s cause an error message and this one isn't reached.

This message could be moved to a target for `Makefile.config` and `Makefile.common` advertising the need to run `configure`, i.e.

```make
Makefile.%:
	@echo "Please refer..."
	@false
```

dra27 reviewed (commented):
I've been wanting to propose this for a while, so I'm thrilled to see it land, @shindere!

My first suggestion about the error message could be moved to another PR if you wished (where it could also be combined with the appropriate test to allow you run `make [dist]clean` without first having to run `configure`).

I'd like to push for the changing of the default target name to `all`, though.

shindere commented:
David Allsopp (2019/09/27 09:14 +0000):
> I've been wanting to propose this for a while, so I'm thrilled to see
> it land, @shindere!

Many thanks for your support, @dra27.

> My first suggestion about the error message could be moved to another
> PR if you wished (where it could also be combined with the appropriate
> test to allow you run `make [dist]clean` without first having to run
> `configure`).

I think you already have some code doing that, right?

> I'd like to push for the changing of the default target name to `all`, though.
>
> > @@ -22,17 +22,7 @@ include Makefile.common
>
>  # For users who don't read the INSTALL file
>  .PHONY: defaultentry
> -defaultentry:
> -ifeq "$(UNIX_OR_WIN32)" "unix"
> -	@echo "Please refer to the installation instructions in file INSTALL."
> -	@echo "If you've just unpacked the distribution, something like"
> -	@echo "	./configure"
> -	@echo "	make world.opt"
> -	@echo "	make install"
> -	@echo "should work.  But see the file INSTALL for more details."
> -else
> -	@echo "Please refer to the instructions in file README.win32.adoc."
> -endif
> +defaultentry: $(DEFAULT_MAKE_TARGET)
>
> The [GNU Coding Standards](https://www.gnu.org/prep/standards/html_node/Standard-Targets.html) recommend this target be called `all`.
>
> At the moment, the `all` target rebuilds the system using whatever is
> presently in `boot/` (so it's used by `world` after first building the
> runtime and `bootstrap` after a bootstrap cycle). It seems to me we
> could safely rename that `all` target and use it here?

I have no strong opinion either way but I think @xavierleroy has,
although I don't know how strong they are, so I'll wait for him to speak
here.

> > @@ -22,17 +22,7 @@ include Makefile.common
>
>  # For users who don't read the INSTALL file
>  .PHONY: defaultentry
> -defaultentry:
> -ifeq "$(UNIX_OR_WIN32)" "unix"
> -	@echo "Please refer to the installation instructions in file INSTALL."
> -	@echo "If you've just unpacked the distribution, something like"
> -	@echo "	./configure"
> -	@echo "	make world.opt"
> -	@echo "	make install"
> -	@echo "should work.  But see the file INSTALL for more details."
> -else
> -	@echo "Please refer to the instructions in file README.win32.adoc."
> -endif
>
> The error message below has been less relevant since 4.08, since if you forget to run `configure`, then these two `include`s cause an error message and this one isn't reached.
>
> This message could be moved to a target for `Makefile.config` and `Makefile.common` advertising the need to run `configure`, i.e.
>
> ```make
> Makefile.%:
> 	@echo "Please refer..."
> 	@false
> ```

Well I seem to remember having read somewhere (perhaps in the GNU coding
standards) that the recommended thing to do is actually to _not_ have a
Makefile by default, but rather a Makefile.in file, so that trying to
run `make` on a not properly configured system just fails. You may argue
that it's less user-friendly, but I think using `./configure` is
well-known enough.

dra27 commented:
The GNU coding standards do indeed recommend having `Makefile.in` if you have `configure` alter any aspect of `Makefile`. That said, given that the Windows used *not* to involve running configure, I think it's slightly more friendly to have a helpful error message for forgetful Windows users, or tired ones who still work on old branches occasionally :wink:

dra27 commented on configure.ac:
```diff
@@ -927,6 +929,10 @@ AS_IF([test x"$enable_native_compiler" = "xno"],
 
+AS_IF([$native_compiler],
+  [default_make_target=world.opt],
+  [default_make_target=world])
```
In the equivalent option in https://github.com/ocaml/ocaml/pull/8835#issuecomment-520249008, you request that the new variables be `true` or `false` (which I'm actually just doing, which is why I noticed).

In this case, it would remove the need for `$default_make_target` completely, as you could do an `ifeq` based solely on `$(NATIVE_COMPILER)` in `Makefile`

shindere commented:
David Allsopp (2019/09/27 08:17 -0700):
> dra27 commented on this pull request.
>
>
>
> > @@ -927,6 +929,10 @@ AS_IF([test x"$enable_native_compiler" = "xno"],
>    AC_MSG_NOTICE([the native compiler is disabled])],
>    [native_compiler=true])
>
> +AS_IF([$native_compiler],
> +  [default_make_target=world.opt],
> +  [default_make_target=world])
>
> In the equivalent option in https://github.com/ocaml/ocaml/pull/8835#issuecomment-520249008, you request that the new variables be `true` or `false` (which I'm actually just doing, which is why I noticed).
>
> In this case, it would remove the need for `$default_make_target`
> completely, as you could do an `ifeq` based solely on
> `$(NATIVE_COMPILER)` in `Makefile`

Thanks. I just rebased and took this comment into account.

dra27 commented:
`BYTECODE_ONLY` was supposed to be a way to instruct the testsuite to skip native code tests … that's actually something I miss. Is there an issue tracking it, before it goes from being an "ignored" option to a "removed" option?

shindere committed:
Let make's default target build the compiler

This commit makes it possible to build the OCaml compiler according to
its configuration by simply runnning make. There is no need to specify
neither world nor world.opt explicitly, although the two targets
remain available.

This commit also introduces (and starts making use of) the
NATIVE_COMPILER build variable whosse value is true when the native
compiler is enabled and false otherwise.

shindere commented:
David Allsopp (2019/09/30 14:00 +0000):
> `BYTECODE_ONLY` was supposed to be a way to instruct the testsuite to
> skip native code tests … that's actually something I miss. Is there an
> issue tracking it, before it goes from being an "ignored" option to a
> "removed" option?

Well normally native tests are naturally skipped now when the native
compiler is not built so I don't think this variable is required. If
there's a need to test whether the native-code compilers are available
explicitly, then I guess `ocamltest` could have an explicit test for
that.

shindere commented:
The error message has now been implemented almost as suggested. Rather
than using `Make.%` I prefered to be explicit and to list the files that
should be regenerated.

dra27 commented:
@shindere - indeed, `BYTECODE_ONLY` was precisely useful to build the native compiler and then *not* test it (see https://github.com/ocaml/ocaml/issues/5199 where it was added, also [Changes#L6520-6521](https://github.com/ocaml/ocaml/blob/trunk/Changes#L6520-L6521)). It used to be that you could say `make -C testsuite BYTECODE_ONLY=true all` whereas at present I think you have to reconfigure and rebuild `ocamltest` (which I literally did last week working on another branch)

dra27 reviewed (approved):
I hope to be able to reach an agreement for the default target change with @xavierleroy, but a decision one way or another doesn't affect this PR. Thanks for the changes, @shindere!

shindere commented:
Thanks a lot, @dra27, will think about the `BYTECODE_ONLY` stuff, not
sure yet about how to go about that one.

shindere commented:
Thanks a lot for the aproval, @dra27.

shindere commented:
David Allsopp (2019/09/30 15:08 +0000):
> @shindere - indeed, `BYTECODE_ONLY` was precisely useful to build the
> native compiler and then *not* test it (see
> https://github.com/ocaml/ocaml/issues/5199 where it was added, also
> [Changes#L6520-6521](https://github.com/ocaml/ocaml/blob/trunk/Changes#L6520-L6521)). It used to be that you could say `make -C testsuite BYTECODE_ONLY=true all` whereas at present I think you have to reconfigure and rebuild `ocamltest` (which I literally did last week working on another branch)

Is this feature useful enogh to start thinking about how it could be
implemented now? Any suggestions re:it's implementation?

gasche commented:
Here is what I observe if I run `make -j3` on a non-configured, clean checkout of 4.10 or trunk:

```
$ make -j3
Please refer to the installation instructions:
Please refer to the installation instructions:
- In file INSTALL for Unix systems.
- In file INSTALL for Unix systems.
- In file README.win32.adoc for Windows systems.
- In file README.win32.adoc for Windows systems.
On Unix systems, if you've just unpacked the distribution,
On Unix systems, if you've just unpacked the distribution,
something like
something like
	./configure
	./configure
	make
	make
	make install
	make install
should work.
make: *** [Makefile:1350: Makefile.config] Error 1
make: *** Waiting for unfinished jobs....
should work.
make: *** [Makefile:1350: Makefile.common] Error 1
```

I suppose that the duplication of each line is not intentional and should be fixed. It appears to come from the parallel build (of course...).

shindere commented:
It would be simpler if we could just not have a `Makefile` as long as
`configure` has not been run. That would solve the problem in what I
feel would be a clean way, rather than trying to hack something
together.

dra27 commented:
That's the automake approach, I think @shindere? We could - although the fix here is really not that hacky, we just need to move the error message to a common dependency - see #9112.

shindere commented:
David Allsopp (2019/11/12 04:55 -0800):
> That's the automake approach, I think @shindere?

Well no I don't think it's that specific. As far as I can remember it
has more to do with projects using autoconf for which it's recommended
that configure creates the Makefile from a Makefile.in script, thus
guaranteeing that make can not be run befor configure.
